src/control/ros/Conversions.cpp

```diff
@@ -0,0 +1,438 @@
+  }
+
+  if (_numCoefficients >= 6)
+  {
+    assert(_currAcceleration.size() == numDofs);
+    assert(_nextAcceleration.size() == numDofs);
+    splineProblem.addConstantConstraint(0, 2, _currAcceleration);
+    splineProblem.addConstantConstraint(1, 2, _nextAcceleration);
+  }
+
+  const auto splineSegment = splineProblem.fit();
+  return splineSegment.getCoefficients()[0];
+}
+
+//=============================================================================
+void extractJointTrajectoryPoint(
+  const trajectory_msgs::JointTrajectory& _trajectory,
+  size_t _index, size_t _numDofs,
+  Eigen::VectorXd* _positions, bool _positionsRequired,
+  Eigen::VectorXd* _velocities, bool _velocitiesRequired,
+  Eigen::VectorXd* _accelerations, bool _accelerationsRequired,
+  std::map<size_t, size_t> indexMap)
+{
+  const auto& waypoint = _trajectory.points[_index];
+
+  try
+  {
+    Eigen::VectorXd trajPos, trajVel, trajAccel;
+    checkVector("positions", waypoint.positions, _numDofs,
+      _positionsRequired, &trajPos);
+    checkVector("velocities", waypoint.velocities, _numDofs,
+      _velocitiesRequired, &trajVel);
+    checkVector("accelerations", waypoint.accelerations, _numDofs,
+      _accelerationsRequired, &trajAccel);
+
+    if ( trajPos.size() > 0 )
+      reorder(indexMap, trajPos, _positions);
+    if ( trajVel.size() > 0 )
+      reorder(indexMap, trajVel, _velocities);
+    if ( trajAccel.size() > 0 )
+      reorder(indexMap, trajAccel, _accelerations);
+  }
+  catch (const std::invalid_argument& e)
+  {
+    std::stringstream message;
+    message << "Waypoint " << _index << " is invalid: " << e.what();
+    throw std::invalid_argument(message.str());
+  }
+}
+
+//=============================================================================
+void extractTrajectoryPoint(
+  const std::shared_ptr<MetaSkeletonStateSpace>& space,
+  const aikido::trajectory::TrajectoryPtr& trajectory,
+  double timeFromStart, const std::map<size_t, size_t>& indexMap,
+  trajectory_msgs::JointTrajectoryPoint& waypoint)
+{
+  const auto numDerivatives = std::min<int>(trajectory->getNumDerivatives(), 1);
+  const auto timeAbsolute = trajectory->getStartTime() + timeFromStart;
+  
+  Eigen::VectorXd tangentVector, jointTangentVector;
+  auto state = space->createState();
+  trajectory->evaluate(timeAbsolute, state);
+  space->logMap(state, tangentVector);
+  assert(tangentVector.size() == numDofs);
+
+  // Reorder tangentVectoraccording to the index mapping
+  reorder(indexMap, tangentVector, &jointTangentVector);
+
+  waypoint.time_from_start = ::ros::Duration(timeFromStart);
+  waypoint.positions.assign(jointTangentVector.data(),
+    jointTangentVector.data() + jointTangentVector.size());
+
+  assert(0 <= numDerivatives && numDerivatives <= 2);
+  const std::array<std::vector<double>*, 2> derivatives{
+    &waypoint.velocities, &waypoint.accelerations};
+  for (int iderivative = 1; iderivative <= numDerivatives; ++iderivative)
```

Comment:
This is the convention I typically use for naming loop variables, so I suspect you're going to have a lot more comments like this on the other `ros_control` pull requests. 😅 

I generally like this convention, but I am fine with @gilwoolee renaming the variables if you insist... 😞